linux: enable Rust (rust-for-linux) in kernel builds (pkgs.linuxKernel.manualConfig)

phip1611 opened the pull request:
###### Problem

I want to compile <https://github.com/Rust-for-Linux/linux> in Nix so that the kernel keeps `CONFIG_RUST=y` and does not silently drop it during the build. The generated `./scripts/target.json` file must also be kept to enable (out-of-tree) module builds. This requires changes in `pkgs.linuxKernel.manualConfig`.

I use the following Nix code to build the kernel. It uses the nixpkgs changes that are also part of this PR.


Nix code:
<details>

```nix
{ lib
, pkgs
, rustc
, rust-bindgen
}:

let
  src =
    let
      rev = "bc22545f38d74473cfef3e9fd65432733435b79f";
    in
    builtins.fetchTarball {
      url = "https://github.com/Rust-for-Linux/linux/archive/${rev}.tar.gz";
      sha256 = "sha256:18mxvk1l97448m9s55zc60nlgwb5dnzl1akpz12xyvigc5kz3q7r";
    };
  # Function that builds a kernel from the provided Linux source with the
  # given config.
  buildKernel = pkgs.linuxKernel.manualConfig {
    inherit (pkgs) stdenv lib;

    inherit src;
    # This is a valid config file with activated Rust support. It works
    # with regular builds (outside of Nix)
    configfile = ./kernel.config;

    # Required so that the kernel activates Rust support. 
    # Uses my local Nix fork at the moment
    # -> https://github.com/NixOS/nixpkgs/pull/232861/files
    extra-build-dependencies = [ rustc rust-bindgen ];

    version = "6.3";
    # Probably that's a weird nixpkgs upstream thingy. Linux 6.3 wants
    # "6.3.0" instead of "6.3".
    modDirVersion = "6.3.0";

    allowImportFromDerivation = true;

    extraMakeFlags = [
      # Helps to detect the output of the rust_is_available.sh script.
      "V=12"
    ];
  };
in
buildKernel
```

</details>

***Edit:*** If `CONFIG_RUST=y` is kept properly, `./scripts/target.json` should also be kept because of

https://github.com/NixOS/nixpkgs/blob/4645ea3569ab7cc4fdc617503f7e54655512a54c/pkgs/os-specific/linux/kernel/manual-config.nix#L363-L364

For example, "rustc" and "rust-bindgen" are required to be present in very specific versions for Linux's early Rust support. Hence, it is handy to optionally pass such deps.

When I build an out-of-tree Rust module like this

<details>

```nix
{ gitignoreSource
, lib
  # The Linux kernel Nix package for which this module is compiled.
, kernel
, stdenv
, rust-bindgen
, rustc
}:

stdenv.mkDerivation rec {
  pname = "qemu-debugcon-driver";
  version = "0.1";

  src = gitignoreSource ../src;

  setSourceRoot = ''
    export sourceRoot=$(pwd)/source
  '';

  nativeBuildInputs = kernel.moduleBuildDependencies
    ++
    # Rust-specific module build depenendencies not specified out-of-the-box in Nix.
    [ rust-bindgen rustc ]
  ;

  makeFlags = kernel.makeFlags ++ [
    "-C"
    "${kernel.dev}/lib/modules/${kernel.modDirVersion}/build"
    "M=$(sourceRoot)"
  ];

  buildFlags = [ "modules" ];
  installFlags = [ "INSTALL_MOD_PATH=${placeholder "out"}" ];
  installTargets = [ "modules_install" ];

  meta = with lib; {
    description = "A QEMU Debugcon Driver for Linux";
    platforms = platforms.linux;
  };
}
```

</details>

It always fails with `target file "./scripts/target.json" does not exist`.

This file is only generated if `CONFIG_RUST=y` and `RUSTC=$(which rustc) BINDGEN=$(which bindgen) scripts/rust_is_available.sh -v` succeeds. ~~I think that with my changes, the kernel build correctly recognizes Rust and that `scripts/rust_is_available.sh` succeeds. However, I'm not sure if `./scripts/target.json` is really generated and if it's kept for the out-of-tree build or is discarded. I suppose that the Nix build environment is already capable of keeping these kind of files.~~ Please look at the discussion as we figured new things out along the way.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

figsoda commented:
@ofborg eval

phip1611 commented:
I really don't know what it takes to make the Kernel build in Nix DO NOT drop CONFIG_RUST=y.
On top of the changes in this PR, I've added `RUSTC=$(which rustc) BINDGEN=$(which bindgen) scripts/rust_is_available.sh -v` to the postPatch hook and could verify that this script succeeds. 

However, the resulting `.config`-file of the kernel build (see #232859) ALWAYS drops all options with `RUST`. If I use a similar build environment directly in a shell without nix-build but just a nix-shell, it works fine...


--
ping @SuperSandro2000 - Could you help me plase?

SuperSandro2000 commented on pkgs/os-specific/linux/kernel/manual-config.nix:
```diff
@@ -50,6 +50,11 @@ in lib.makeOverridable ({
   # ignored
   features ? null, lib ? lib_, stdenv ? stdenv_,
+
+  # Additional build dependencies. For example, "rustc" and "rust-bindgen" are
+  # required to be present in very specific versions for Linux's early Rust
+  # support. Hence, it is handy to optionally pass such deps.
+  extra-build-dependencies ? []
```
```suggestion
  # Additional native build dependencies.
  # For example, "rustc" and "rust-bindgen" are required to be present in very specific versions for Linux's early Rust support.
  extraNativeBuildInputs ? []
```

SuperSandro2000 commented on pkgs/os-specific/linux/kernel/manual-config.nix:
```diff
@@ -67,7 +72,8 @@ let
     (buildPackages.deterministic-uname.override { inherit modDirVersion; })
-  ] ++ optional (lib.versionAtLeast version "5.13") zstd;
+  ] ++ optional (lib.versionAtLeast version "5.13") zstd
+  ++ extra-build-dependencies;
```
```suggestion
  ++ extraNativeBuildInputs;
```

SuperSandro2000 commented on pkgs/os-specific/linux/kernel/manual-config.nix:
```diff
@@ -106,6 +112,7 @@ stdenv.mkDerivation ({
       ++ optionals (lib.versionAtLeast version "4.16") [ bison flex ]
       ++ optionals (lib.versionAtLeast version "5.2")  [ cpio pahole zlib ]
       ++ optional  (lib.versionAtLeast version "5.8")  elfutils
+      ++ extra-build-dependencies
```
```suggestion
      ++ extraNativeBuildInputs
```

alyssais reviewed (changes requested):
This doesn't "enable Rust" at all.  It just adds a redundant way for a caller to override dependencies.  Packages should know their own dependencies, not force the caller to decide what they are.  If, in some special situation, the caller does need to override dependencies, they can already do that with `overrideAttrs`.

If you want to add Rust support for kernels to Nixpkgs, just add the dependencies normally.  You could gate it behind the existing `features` parameter to avoid adding the dependencies to kernels without useful Rust code in them, and then we can enable it by default once there are actual user-observable features that need Rust.

alyssais commented:
No, I mean that the expression in manual-config.nix should add the Rust dependencies itself, like it already does for every other dependency of the kernel.

phip1611 commented:
How will it then be possible to pass in very specific versions of rustc and rust-bindgen except for using an nixpkgs overlay?

alyssais commented:
Well, we have the same `override` mechanism available that every other package has.

> Again, the kernel build is very picky about specific versions of those tools. At least for the early Rust support. Will be more relaxed in a year or so.

Is it actually? The kernel documentation says

> Newer versions may or may not work because the kernel depends on some unstable Rust features, for the moment.

Is this actually a problem in practice?  Are there any problems when using the versions of Rust and bindgen in Nixpkgs?

phip1611 commented:
> Is this actually a problem in practice? Are there any problems when using the versions of Rust and bindgen in Nixpkgs?

Yes, it is. For example, with the latest Rust-for-Linux state and the latest nixpkgs state, the build fails as rust-bindgen is too new. However, if I downgrade rust-bindgen to an older version, (0.61 instead of 0.63), it works.

alyssais commented:
AIUI Rust drivers are likely to start showing up in 6.5 or 6.6, so unless the bindgen version requirements are going to relax _very_ soon, we're going to want Rust enabled by default even with the requirements for special versions.  There's not a lot of time until then, and I don't expect much to change, so we might as well start working towards having Rust enabled by default now.  There's no point making everybody that wants to use it write the same code out of tree.

We should revisit https://github.com/NixOS/nixpkgs/pull/179581 and introduce a package for whatever bindgen version is currently required by Linux, and then have the Linux build enable Rust by default.

alyssais commented:
(I've changed my mind somewhat about the `features` parameter — I don't think there's much reason not to just enable Rust by default.)

phip1611 commented:
At the moment, I'm still fighting the problem that `manualConfig` always discards `CONFIG_RUST=y`, but only for nixpkgs/master. On nixpkgs/22.11. this doesn't happen. Will update the PR soon accordingly.

phip1611 committed:
XXX wip: linux: enable Rust (rust-for-linux) in kernel builds

This commit enables `pkgs.linuxKernel.manualConfig` to build kernels that have
set CONFIG_RUST=y. For this, it requries rustc and rust-bindgen in path.

To debug problems such as that the kernel build silently removes the
CONFIG_RUST=y option, it is useful to add `make LLVM=1 rustavailable && exit 1`
to the postPatch step and view the result.

phip1611 commented:
> One issue that this PR still has is that bindgen builds against the wrong clang version.

To be precisely: Also passing in a valid `rust-bindgen` from the outside doesn't work. `make rustavailable` works in the `postPatch` step but fails in the `configurePhase` step. It always uses the internal bindgen.

The line `make $makeFlags "''${makeFlagsArray[@]}" oldconfig` is the reason where `make $makeFlags "''${makeFlagsArray[@]}" rustavailable` fails.

I don't know the exact details but providing a specific `kernel-rust-bindgen` within nixpkgs will be the right thing to do.

phip1611 commented:
Closing in favor of https://github.com/NixOS/nixpkgs/pull/234311
